### PR TITLE
[MI-2049]: Added websocket support to detect user connection details …

### PR DESCRIPTION
**Pull request by avas27JTG**
Added WebSocket support to detect user connection details and a centralized check for root modals

### CODE REVIEW

File: server/serializers/oAuth.go [hunk 1]
```diff
@@ -18,3 +18,7 @@ type OAuthSuccessResponse struct {
 	RefreshToken string `json:"refresh_token"`
 	ExpiresIn    string `json:"expires_in"`
 }
+
+type ConnectedResponse struct {
+	Connected bool `json:"connected"`
```
Comment (Kshitij-Katiyar) [inline server/serializers/oAuth.go:23]:
```suggestion
	IsConnected bool `json:"connected"`
```

**Review by Kshitij-Katiyar — APPROVED** [review]

**Review by sooraj-shukla — APPROVED** [review]

File: webapp/src/reducers/linkModal/index.ts [hunk 2]
```diff
@@ -13,28 +13,27 @@ export const openLinkModalSlice = createSlice({
...
+            state.isLinked = false;
+
+            if (action.payload.commandArgs.length > 0) {
+                const {organization, project} = getProjectLinkModalArgs(action.payload.commandArgs[0]) as LinkPayload;
+                state.organization = organization;
+                state.project = project;
+            }
+        },
+        toggleIsLinked: (state: LinkProjectModalState, action: PayloadAction<boolean>) => {
+            state.isLinked = action.payload;
```
Comment (manojmalik20) [inline webapp/src/reducers/linkModal/index.ts:29]: Why is this repeated here?

File: webapp/src/selectors/index.tsx [hunk 3]
```diff
@@ -6,6 +6,14 @@ export const getRhsState = (state: ReduxState): {isSidebarOpen: boolean} => {
     return state.views.rhs;
 };
 
+export const getGlobalModalState = (state: ReduxState): GlobalModalState => {
+    return state['plugins-mattermost-plugin-azure-devops'].globalModalSlice;
```
Comment (manojmalik20) [inline webapp/src/selectors/index.tsx:10]: Can't you use a constant here for the plugin id?

File: webapp/src/hooks/usePluginApi.ts [hunk 4]
```diff
@@ -18,7 +18,11 @@ function usePluginApi() {
         return {data, isError, isLoading, isSuccess};
     };
 
-    return {makeApiRequest, getApiState, state};
+    const isUserAccountConnected = (): boolean => {
+        return state['plugins-mattermost-plugin-azure-devops'].userConnectedSlice.isConnected;
```
[the remark this answers is not in the document]
Comment (avas27JTG) [reply webapp/src/hooks/usePluginApi.ts:22]: Okay, will refactor this in later PR as this is old PR doing here will not update all

File: webapp/src/reducers/linkModal/index.ts (hunk 2 above)
Comment (avas27JTG) [reply webapp/src/reducers/linkModal/index.ts:29]: it was due to sync

File: webapp/src/selectors/index.tsx (hunk 3 above)
Comment (avas27JTG) [reply webapp/src/selectors/index.tsx:10]: this is already a kind of constant as we have defined this as a type in `ReduxState` so it's a string literal and TS will not allow any value other than this

File: webapp/src/types/common/store.d.ts [hunk 5]
```diff
@@ -1,6 +1,29 @@
+interface ReduxState extends GlobalState {
+    views: {
+        rhs: {
+            isSidebarOpen: boolean
+        }
+    }
+    'plugins-mattermost-plugin-azure-devops': RootState<{ [x: string]: QueryDefinition<void, BaseQueryFn<string | FetchArgs, unknown, FetchBaseQueryError, {}, FetchBaseQueryMeta>, never, WellList[], 'pluginApi'>; }, never, 'pluginApi'>
+}
```
[the remark this answers is not in the document]
Comment (avas27JTG) [reply webapp/src/types/common/store.d.ts:8]: okay will do this in later or most recent PR/branch

File: webapp/src/containers/Rhs/index.tsx [hunk 6]
```diff
@@ -1,44 +1,25 @@
...
-            {isLoading && <LinearLoader/>}
-            {isError && <AccountNotLinked/>}
             {
-                !isLoading &&
-                !isError &&
-                isSuccess && (
+                !isUserAccountConnected() && <AccountNotLinked/>
+            }
+            {
+                isUserAccountConnected() && (
```
[the remark this answers is not in the document]
Comment (avas27JTG) [reply webapp/src/containers/Rhs/index.tsx:19]: I tried this but eslint gave error, also if we flip the condition we still need nested ternary operator so this way or other we will end up doing similar thing
image

File: webapp/src/websocket/index.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import {Store, Action} from 'redux';
+import {GlobalState} from 'mattermost-redux/types/store';
+
+import {toggleIsConnected} from 'reducers/userConnected';
+
+export function handleConnect(store: Store<GlobalState, Action<Record<string, unknown>>>) {
+    return (_: any) => {
```
Comment (manojmalik20) [inline webapp/src/websocket/index.ts:7]:
```suggestion
    return (_: WebsocketEventParams) => {
```

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**